inc/class-wpseo-endpoint-factory.php

```diff
@@ -0,0 +1,173 @@
+	 */
+	private $method;
+
+	/**
+	 * WPSEO_Endpoint_Factory constructor.
+	 *
+	 * @param string   $namespace           The endpoint's namespace.
+	 * @param string   $endpoint            The endpoint's URL.
+	 * @param callable $callback            The callback function to execute.
+	 * @param callable $permission_callback The permission callback to execute to determine permissions.
+	 * @param string   $method              The HTTP method to use. Defaults to GET.
+	 *
+	 * @throws WPSEO_Invalid_Argument_Exception The invalid argument exception.
+	 */
+	public function __construct( $namespace, $endpoint, $callback, $permission_callback, $method = 'GET' ) {
+		if ( ! WPSEO_Validator::is_string( $namespace ) ) {
+			throw WPSEO_Invalid_Argument_Exception::invalid_string_parameter( $namespace, 'namespace' );
+		}
+
+		$this->namespace = $namespace;
+
+		if ( ! WPSEO_Validator::is_string( $endpoint ) ) {
+			throw WPSEO_Invalid_Argument_Exception::invalid_string_parameter( $endpoint, 'endpoint' );
+		}
+
+		$this->endpoint = $endpoint;
+
+		if ( ! is_callable( $callback ) ) {
+			throw WPSEO_Invalid_Argument_Exception::invalid_callable_parameter( $callback, 'callback' );
+		}
+
+		$this->callback = $callback;
+
+		if ( ! is_callable( $permission_callback ) ) {
+			throw WPSEO_Invalid_Argument_Exception::invalid_callable_parameter( $permission_callback, 'callback' );
+		}
+
+		$this->permission_callback = $permission_callback;
+
+		$this->method = $this->validate_method( $method );
+	}
+
+	/**
+	 * Gets the associated arguments.
+	 *
+	 * @return array The arguments.
+	 */
+	public function get_arguments() {
+		return $this->args;
+	}
+
+	/**
+	 * Determines whether or not there are any arguments present.
+	 *
+	 * @return bool Whether or not any arguments are present.
+	 */
+	public function has_arguments() {
+		return count( $this->args ) > 0;
+	}
+
+	/**
+	 * Registers the endpoint with WordPress.
+	 *
+	 * @return void
+	 */
+	public function register() {
+		$config = array(
+			'methods'			  => $this->method,
+			'callback'			  => $this->callback,
+			'permission_callback' => $this->permission_callback,
+		);
+
+		if ( $this->has_arguments() ) {
+			$config['args'] = $this->args;
+		}
+
+		register_rest_route( $this->namespace, $this->endpoint, $config );
+	}
+
+	/**
+	 * Validates the method parameter.
+	 *
+	 * @param string $method The set method parameter.
+	 *
+	 * @return string The validated method.
+	 *
+	 * @throws WPSEO_Invalid_Argument_Exception The invalid argument exception.
+	 * @throws InvalidArgumentException The invalid argument exception.
+	 */
+	protected function validate_method( $method ) {
+		if ( ! WPSEO_Validator::is_string( $method ) ) {
+			throw WPSEO_Invalid_Argument_Exception::invalid_string_parameter( $method, 'method' );
+		}
+
+		if ( ! in_array( $method, $this->valid_http_methods, true ) ) {
+			throw new InvalidArgumentException( sprintf( '%s is not a valid HTTP method', $method ) );
```

Comment:
Core supports multiple methods for a certain endpoint, partly for compatibility. Since we allow editing an indexable, I think it should use `POST, PUT, PATCH` for methods, in parity with core.

Reply:
This can be ignored, per other comment discussion we figured out it should remain `PATCH` only.